server/schedule/operator/step.go

```diff
@@ -437,3 +449,323 @@ func (al AddLightLearner) Influence(opInfluence OpInfluence, region *core.Region
+		_, _ = fmt.Fprintf(b, ", demote voter peer %v on store %v to learner", dv.PeerID, dv.ToStore)
+	}
+	return b.String()
+}
+
+// ConfVerChanged returns how much the conf version will increase due to this step.
+func (cpe ChangePeerV2Enter) ConfVerChanged(region *core.RegionInfo) uint64 {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer == nil || peer.Id != pl.PeerID ||
+			(peer.Role != metapb.PeerRole_IncomingVoter && peer.Role != metapb.PeerRole_Voter) {
+			return 0
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStoreVoter(dv.ToStore)
+		if peer != nil && (peer.Id != dv.PeerID || (peer.Role != metapb.PeerRole_DemotingVoter && peer.Role != metapb.PeerRole_Learner)) {
+			return 0
+		}
+	}
+	return uint64(len(cpe.PromoteLearners) + len(cpe.DemoteVoters))
+}
+
+// IsFinish checks if current step is finished.
+func (cpe ChangePeerV2Enter) IsFinish(region *core.RegionInfo) bool {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer != nil && peer.Id != pl.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", pl.String()), zap.Uint64("obtain-voter", peer.Id))
+		}
+		if peer == nil || peer.Id != pl.PeerID || peer.Role != metapb.PeerRole_IncomingVoter {
+			return false
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStoreVoter(dv.ToStore)
+		if peer != nil && peer.Id != dv.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", dv.String()), zap.Uint64("obtain-learner", peer.Id))
+		}
+		if peer == nil || peer.Id != dv.PeerID || peer.Role != metapb.PeerRole_DemotingVoter {
+			return false
+		}
+	}
+	return true
+}
+
+// CheckSafety checks if the step meets the safety properties.
+func (cpe ChangePeerV2Enter) CheckSafety(region *core.RegionInfo) error {
+	inJointState, notInJointState := false, false
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStorePeer(pl.ToStore)
+		if peer.GetId() != pl.PeerID {
+			return errors.New("peer does not exist")
+		}
+		switch peer.Role {
+		case metapb.PeerRole_Learner:
+			notInJointState = true
+		case metapb.PeerRole_IncomingVoter:
+			inJointState = true
+		default:
+			return errors.New("unexpected peer role")
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStorePeer(dv.ToStore)
+		if peer.GetId() != dv.PeerID {
+			return errors.New("peer does not exist")
+		}
+		switch peer.Role {
+		case metapb.PeerRole_Voter:
+			notInJointState = true
+		case metapb.PeerRole_DemotingVoter:
+			inJointState = true
+		default:
+			return errors.New("unexpected peer role")
+		}
+	}
+
+	count := core.CountInJointState(region.GetPeers()...)
+	if (notInJointState && inJointState) ||
+		(notInJointState && count != 0) ||
+		(inJointState && count != len(cpe.PromoteLearners)+len(cpe.DemoteVoters)) {
+		// change is not atomic, or there are other peers in the joint state
+		return errors.New("unexpected peer role")
```

Comment:
better to add more detail, or we cannot know which condition fail

Reply:
Different error messages have been returned for different situations. PTAL